### ISSUE TITLE
forbidden attributes

### ISSUE THREAD

**Issue by lindem** [ember-template-lint/ember-template-lint#427]
Testing the waters: We would like to have a rule to ember-template-lint that is configurable via a string array and warns of components that use attributes only development versions should use (mainly, developing helpers).

Is there interest to merge it here once it's done? 

It is going to be called `no-forbidden-attributes-in-components` (maybe somebody has a better name?)

**Comment by rwjblue** [comment]
Hmm, I am unsure. There is a recently added [invocable-blacklist](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/invocable-blacklist.md) rule, is that what you are thinking of (it would allow blacklisting helpers/keywords/components that are in the list)?

I think if we did do this rule, it would have to be per-component as each component has different APIs and its pretty unlikely that you just want to completely forbid a given argument name to a component...

**Comment by rwjblue** [comment]
I'm happy to work/iterate in a PR with you on it, I'm just not sure how broadly useful it is...

**Comment by lindem** [comment]
The situation is that we support an "experimental" flag on a lot of our components and want to not have usage of this in our production builds. I can imagine that a wildcard is probably something we could consider.

But the idea to have a list per-component is probably a better idea than what I had.

> Hmm, I am unsure. There is a recently added invocable-blacklist rule, is that what you are thinking of (it would allow blacklisting helpers/keywords/components that are in the list)?

Yes, precisely. 

What about the following: 

```javascript
// ...
forbidden-attributes: {
  "*": ["experimental"],
  "another-component": [/* ... some attrs here ...*/]
} 
```

The rule would then apply things in the wildcard to every component?

**Comment by lindem** [comment]
on second thought, the invocable-blacklist can probably be used for this; the rule covers invocations that are not to be used, after all.

I just don't know how to sensibly extend it to not break what's already there.

How about allowing configuration as follows: 

```
invocable-blacklist: [
  // never allowed to use this!
  'forbidden-helper', 
  // this next one is allowed, unless it has some attribute
  {
    name: 'perfectly-fine-component',
    withAttributes: ['forbiddenAttribute']
  }, 
  // this next one prohibits an attribute or named parameter
  // for all components
  {
    name: '*',
    withAttributes: ['experimental']
  }
]
```

Changing it in that way might have advantages:
* it does not introduce a new rule name
* it lets us blacklist more precisely (and it is extensibly future-proof as well)
* ...?

The disadvantages are 
* too much configuration?
* ...others?

Thoughts?

**Comment by lindem** [comment]
Just wanted to check back: I got swamped with (other, but also ember-related) work. I am still on this (it is a tracked issue at our company as well) and I will get back to it. If it turns out that we come up with another solution, I will say so here and close accordingly.

**Comment by lindem** [comment]
through ongoing issues on our side the requirements (in our project using ember-template-lint) have changed in such a way that I think it's safe to say that we are going to implement a rule on our side which will cover our immediate needs, but is probably not useful for a broader audience.

However, should it turn out to be a more generally useful rule in the end, I will be back with it.

Thank you for your patience while letting me sort this out. Closing.
